cloudinit/distros/alpine.py

```diff
@@ -202,6 +204,423 @@ def preferred_ntp_clients(self):
 
         return self._preferred_ntp_clients
 
+    def add_user(self, name, **kwargs):
+        """
+        Add a user to the system using standard tools
+
+        On Alpine this may use either useradd or adduser depending on whether
+        the shadow package is installed.
+        """
+        if util.is_user(name):
+            LOG.info("User %s already exists, skipping.", name)
+            return
+
+        if "create_groups" in kwargs:
+            create_groups = kwargs.pop("create_groups")
+        else:
+            create_groups = True
+
+        # Decide whether 'useradd' or 'adduser' will be used
+        if subp.which("useradd"):
+            # Shadow useradd
+            use_useradd = True
+        else:
+            # Busybox adduser
+            use_useradd = False
+
+        # Since we are creating users, we want to carefully validate the
+        # inputs. If something goes wrong, we can end up with a system
+        # that nobody can login to.
+        if use_useradd:
+            # Shadow package is installed, use 'useradd'
+            add_cmd = ["useradd", name]
+            log_add_cmd = ["useradd", name]
+
+            add_opts = {
+                "expiredate": "--expiredate",
+                "gecos": "--comment",
+                "groups": "--groups",
+                "homedir": "--home",
+                "inactive": "--inactive",
+                "passwd": "--password",
+                "primary_group": "--gid",
+                "shell": "--shell",
+                "uid": "--uid",
+            }
+
+            add_flags = {
+                "no_log_init": "--no-log-init",
+                "no_user_group": "--no-user-group",
+                "system": "--system",
+            }
+        else:
+            # Shadow package is not installed, use Busybox 'adduser'
+            add_cmd = ["adduser", "-D"]
+
+            add_opts = {
+                "expiredate": "",
+                "gecos": "-g",
+                "groups": "",
+                "homedir": "-h",
+                "inactive": "",
+                "passwd": "",
+                "primary_group": "-G",
+                "shell": "-s",
+                "uid": "-u",
+            }
+
+            add_flags = {"system": "-S"}
+
+        if use_useradd:
+            redact_opts = ["passwd"]
+
+        # support kwargs having groups=[list] or groups="g1,g2"
+        groups = kwargs.get("groups")
+        if groups:
+            if isinstance(groups, str):
+                groups = groups.split(",")
+
+            if isinstance(groups, dict):
+                LOG.warning(
```

Comment:
Why not use `util.deprecate()` here?

Reply:
I've had this code sitting around locally from before util.deprecate() was added in Feb 2013. I thought it was about time to upstream this :-)

I'll change it to use that now.

Reply:
Done.